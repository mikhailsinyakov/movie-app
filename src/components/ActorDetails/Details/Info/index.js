import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { useTranslation } from "react-i18next";
import withInfoStyles from "hocs/withInfoStyles";
import Description from "./Description";

const formatDate = (date, language) => {
  if (language === "en") return date;
  const [year, month, day] = date.split("-");
  return `${day}.${month}.${year}`;
}

const StyledName = styled.p`
  text-align: center;
  font-weight: bold;
`;

const Info = ({
  name,
  birthday,
  place_of_birth,
  biography,
  className
}) => {
  const { i18n : {language} } = useTranslation();
  
  return (
    <div className={className}>
      <StyledName>{name}</StyledName>
      <Description
        biography={biography}
        birthday={birthday && formatDate(birthday, language)}
        place_of_birth={place_of_birth}
      />
    </div>
  );
};

Info.propTypes = {
  name: PropTypes.string.isRequired,
  birthday: PropTypes.string,
  place_of_birth: PropTypes.string,
  biography: PropTypes.string.isRequired
};

export default withInfoStyles(Info);
