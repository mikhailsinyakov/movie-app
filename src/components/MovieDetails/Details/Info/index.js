import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Title from "./Title";
import Description from "./Description";

const formatTime = minOverall => {
  const hours = Math.floor(minOverall / 60);
  const minutes = (minOverall % 60).toString().padStart(2, "0");

  return `${hours}:${minutes}`;
};

const Info = ({
  title,
  original_title,
  vote_average,
  overview,
  genres,
  release_date,
  production_countries,
  runtime,
  budget,
  revenue,
  actors,
  className
}) => (
  <div className={className}>
    <Title title={title} original_title={original_title} />
    <Description
      overview={overview}
      genres={genres}
      rating={vote_average ? vote_average.toFixed(1) : null}
      release={+release_date.split("-")[0]}
      runtime={runtime ? formatTime(runtime) : ""}
      production={
        production_countries.length ? production_countries.join(", ") : ""
      }
      budget={budget ? "$" + (budget / 1000000).toFixed(1) + "M" : ""}
      revenue={revenue ? "$" + (revenue / 1000000).toFixed(1) + "M" : ""}
      actors={actors}
    />
  </div>
);

Info.propTypes = {
  title: PropTypes.string.isRequired,
  original_title: PropTypes.string.isRequired,
  vote_average: PropTypes.number.isRequired,
  overview: PropTypes.string,
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  release_date: PropTypes.string.isRequired,
  production_countries: PropTypes.arrayOf(PropTypes.string).isRequired,
  runtime: PropTypes.number,
  budget: PropTypes.number,
  revenue: PropTypes.number,
  actors: PropTypes.array.isRequired,
};

const StyledInfo = styled(Info)`
  max-width: 330px;
  background-color: rgba(226, 243, 187, 0.7);
  padding: 1rem;
  margin: 1rem;
  border-radius: 0.5rem;

  @media screen and (min-width: 550px) {
    & {
      width: 250px;
      margin-top: 0;
      margin-left: 2rem;
      margin-right: 2rem;
      padding-left: 2rem;
      padding-right: 2rem;
    }
  }

  @media screen and (min-width: 800px) {
    & {
      width: 600px;
    }
  }
`;

export default StyledInfo;
