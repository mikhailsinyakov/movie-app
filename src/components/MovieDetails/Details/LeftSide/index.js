import React from "react";
import PropTypes from "prop-types";
import Poster from "../../../../shared/Poster";
import Money from "./Money";

const LeftSide = ({ title, poster_src, budget, revenue }) => (
  <div>
    <Poster source={poster_src} title={title} />
    <Money budget={budget} revenue={revenue} />
  </div>
);

LeftSide.propTypes = {
  title: PropTypes.string.isRequired,
  poster_src: PropTypes.string.isRequired,
  budget: PropTypes.number.isRequired,
  revenue: PropTypes.number.isRequired
};

export default LeftSide;
