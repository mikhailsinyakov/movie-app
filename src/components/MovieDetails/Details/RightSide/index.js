import React from "react";
import PropTypes from "prop-types";
import Title from "./Title";
import Description from "./Description";

const RightSide = ({
  title,
  original_title,
  vote_average,
  overview,
  release_date,
  production_countries,
  runtime
}) => (
  <div>
    <Title
      title={title}
      original_title={original_title}
      rating={vote_average}
    />
    <Description
      overview={overview}
      release={+release_date.split("-")[0]}
      production={production_countries}
      runtime={runtime && `${Math.floor(runtime / 60)}:${runtime % 60}`}
    />
  </div>
);

RightSide.propTypes = {
  title: PropTypes.string.isRequired,
  original_title: PropTypes.string.isRequired,
  vote_average: PropTypes.number.isRequired,
  overview: PropTypes.string,
  release_date: PropTypes.string.isRequired,
  production_countries: PropTypes.arrayOf(PropTypes.string).isRequired,
  runtime: PropTypes.number
};

export default RightSide;
