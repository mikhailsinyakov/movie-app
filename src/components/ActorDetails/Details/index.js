import React, { useRef } from "react";
import PropTypes from "prop-types";
import withDetailsStyles from "hocs/withDetailsStyles";
import withImageStyles from "hocs/withImageStyles";
import useScrollToMoveElement from "hooks/useScrollToMoveElement";
import Poster from "shared/Poster";
import Info from "./Info";

const StyledPhoto = withImageStyles(Poster);

const Details = ({ details, className }) => {
  const posterEl = useRef();
  useScrollToMoveElement(posterEl);

  return (
    <div className={className}>
      <StyledPhoto 
        source={details.profile_src} 
        alt={details.name} 
        key={details.profile_src}
        ref={posterEl}
       />
      <Info {...details} />
    </div>
  );
}

Details.propTypes = {
  details: PropTypes.shape({
    name: PropTypes.string.isRequired,
    profile_src: PropTypes.string,
    birthday: PropTypes.string,
    place_of_birth: PropTypes.string,
    biography: PropTypes.string.isRequired
  })
};

export default withDetailsStyles(Details);
