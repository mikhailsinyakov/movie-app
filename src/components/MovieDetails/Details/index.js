import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Poster from "shared/Poster";
import Info from "./Info";

const StyledPoster = styled(Poster)`
  align-self: flex-start;
  border-bottom-left-radius: 1rem;
  border-bottom-right-radius: 1rem;
  
  @media screen and (min-width: 600px) {
    position: fixed;
    left: calc((100vw - 600px) * 0.6);
  }

  @media screen and (min-width: 700px) {
    left: calc((100vw - 700px) * 0.3 + 60px);
    width: 250px;
  }
`;

const Details = ({ details, className }) => (
  <div className={className}>
    <StyledPoster 
      source={details.poster_src} 
      title={details.title} 
      key={details.poster_src}
     />
    <Info {...details} />
  </div>
);

Details.propTypes = {
  details: PropTypes.shape({
    title: PropTypes.string.isRequired,
    poster_src: PropTypes.string,
    budget: PropTypes.number.isRequired,
    revenue: PropTypes.number.isRequired,
    original_title: PropTypes.string.isRequired,
    vote_average: PropTypes.number.isRequired,
    overview: PropTypes.string,
    release_date: PropTypes.string.isRequired,
    production_countries: PropTypes.arrayOf(PropTypes.string).isRequired,
    runtime: PropTypes.number,
    genres: PropTypes.arrayOf(PropTypes.string).isRequired
  })
};

const StyledDetails = styled(Details)`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  margin-top: 1rem;
  
  @media screen and (min-width: 600px) {
    margin-left: calc((100vw - 600px) * 0.6 + 200px);
    justify-content: flex-start;
  }
  
  @media screen and (min-width: 700px) {
    margin-left: calc((100vw - 700px) * 0.3 + 310px);
  }
`;

export default StyledDetails;
