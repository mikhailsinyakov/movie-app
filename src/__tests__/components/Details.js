import React from "react";
import Renderer from "react-test-renderer";
import { render, getByText } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import withLanguage from "../../components/Language";
import Details from "../../components/MovieDetails/Details";

const Parent = ({ children }) => <div>{children}</div>;
const ParentWithLanguage = withLanguage(Parent);

const details = {
  budget: 1500,
  genres: ["Fantasy"],
  original_title: "Dragonheart: Vengeance original",
  overview:
    "Lukas, a young farmer whose family is killed by savage raiders in the countryside, sets out on an epic quest for revenge, forming an unlikely trio with a majestic dragon and a swashbuckling, sword-fighting mercenary, Darius.",
  poster_src: "https://image.tmdb.org/t/p/w342/qZ1KAgfdeNbzrNYKW4BIRHdEBJ9.jpg",
  production_countries: ["United States of America"],
  release_date: "2020-02-04",
  revenue: 1000,
  runtime: 97,
  title: "Dragonheart: Vengeance",
  vote_average: 6
};

describe("Snapshot", () => {
  let component;
  beforeEach(() => {
    component = Renderer.create(
      <ParentWithLanguage>
        <Details details={details} />
      </ParentWithLanguage>
    );
  });
  afterEach(() => {
    component.unmount();
  });
  it("matches snapshot", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});

const renderComponent = () =>
  render(
    <ParentWithLanguage>
      <Details details={details} />
    </ParentWithLanguage>
  );

describe("Details", () => {
  it("renders an image", () => {
    const { getByAltText } = renderComponent();
    expect(getByAltText(details.title)).toBeInTheDocument();
  });

  it("renders title", () => {
    const { getByText } = renderComponent();
    expect(getByText(new RegExp(details.title))).toBeInTheDocument();
    expect(getByText(new RegExp(details.original_title))).toBeInTheDocument();
  });

  it("renders overview", () => {
    const { getByText } = renderComponent();
    expect(getByText(details.overview)).toBeInTheDocument();
  });

  it("renders release year", () => {
    const { getByText } = renderComponent();
    const release_year = details.release_date.split("-")[0];
    expect(getByText(new RegExp(release_year))).toBeInTheDocument();
  });

  it("renders rating", () => {
    const { getByText } = renderComponent();
    expect(getByText(details.vote_average.toFixed(1))).toBeInTheDocument();
  });

  it("renders production country", () => {
    const { getByText } = renderComponent();
    expect(
      getByText(new RegExp(details.production_countries[0]))
    ).toBeInTheDocument();
  });

  it("renders runtime", () => {
    const { getByText } = renderComponent();
    const runtime =
      Math.floor(details.runtime / 60) + ":" + (details.runtime % 60);
    expect(getByText(new RegExp(runtime))).toBeInTheDocument();
  });

  it("renders budget", () => {
    const { getByText } = renderComponent();
    expect(getByText(new RegExp(details.budget))).toBeInTheDocument();
  });

  it("renders revenue", () => {
    const { getByText } = renderComponent();
    expect(getByText(new RegExp(details.revenue))).toBeInTheDocument();
  });
});
