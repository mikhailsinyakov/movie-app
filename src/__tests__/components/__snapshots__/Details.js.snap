// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot matches snapshot 1`] = `
<div>
  <div
    className="sc-fzplWN iJnEoU"
  >
    <img
      alt="Dragonheart: Vengeance"
      className="sc-AxjAm sc-fzpans eqmmmZ"
      onError={[Function]}
      onLoad={[Function]}
      src="img/poster-placeholder.png"
    />
    <div
      className="sc-fzoLsD gDwgRa"
    >
      <p
        className="sc-AxiKw keojvp"
      >
        Dragonheart: Vengeance
         
        <span
          className="sc-AxirZ jUXLfU"
        >
          (
          Дракон
          )
        </span>
      </p>
      <div>
        <p
          className="sc-AxmLO dgSOwn"
        >
          Lukas, a young farmer whose family is killed by savage raiders in the countryside, sets out on an epic quest for revenge, forming an unlikely trio with a majestic dragon and a swashbuckling, sword-fighting mercenary, Darius.
        </p>
        <p
          className="sc-AxhUy sc-fzozJi iWFcxS"
        >
          <span
            className="sc-AxhCb iA-dFqF"
          >
            Fantasy
          </span>
        </p>
        <p
          className="sc-Axmtr clcwvu"
        >
          <span
            className="sc-AxgMl kvVMrz"
          >
            Rating
            : 
          </span>
           
          <span
            className="sc-AxheI fqiWrt"
          >
            6.0
          </span>
        </p>
        <p
          className="sc-Axmtr clcwvu"
        >
          <span
            className="sc-AxgMl kvVMrz"
          >
            Release year
            : 
          </span>
           
          <span
            className="sc-AxheI fqiWrt"
          >
            2020
          </span>
        </p>
        <p
          className="sc-Axmtr clcwvu"
        >
          <span
            className="sc-AxgMl kvVMrz"
          >
            Runtime
            : 
          </span>
           
          <span
            className="sc-AxheI fqiWrt"
          >
            1:37
          </span>
        </p>
        <p
          className="sc-Axmtr dShpiR"
        >
          <span
            className="sc-AxgMl kvVMrz"
          >
            Production
            : 
          </span>
           
          <span
            className="sc-AxheI fqiWrt"
          >
            United States of America
          </span>
        </p>
        <p
          className="sc-Axmtr dShpiR"
        >
          <span
            className="sc-AxgMl kvVMrz"
          >
            Budget
            : 
          </span>
           
          <span
            className="sc-AxheI fqiWrt"
          >
            $15.0M
          </span>
        </p>
        <p
          className="sc-Axmtr dShpiR"
        >
          <span
            className="sc-AxgMl kvVMrz"
          >
            Revenue
            : 
          </span>
           
          <span
            className="sc-AxheI fqiWrt"
          >
            $10.0M
          </span>
        </p>
      </div>
    </div>
  </div>
</div>
`;
